### PR TITLE
HPCC-18132 - ESDL support include paths for searching EspInclude files

### PR DESCRIPTION
**Pull request by mayx**
- Add the -i --includePath options to relevant esdl commands
- Add support in esdl compiler to search included files in the include
  paths

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by mayx** [comment]
@afishbeck please review...

File: tools/esdlcmd-xml/esdl2xml.cpp [hunk 1]
```diff
@@ -33,6 +35,7 @@ void static usage(const char* programName)
 
     puts("Available options:");
     puts(" -r|--recursive: process all includes");
+    puts(" -i|--includePath <include path>: Locations to look for included esdl files");
```
Comment (afishbeck) [inline tools/esdlcmd-xml/esdl2xml.cpp]: eclcc and eclcmd use -I maybe we should be consistent?

File: tools/esdlcmd-xml/esdl2xml.cpp [hunk 2]
```diff
@@ -75,13 +83,37 @@ void parseCommandLine(int argc, char* argv[], Esdl2Esxdl * cmd)
         usage(argv[0]);
 }
 
+bool extractEsdlxmlCmdOption(StringBuffer & option, IProperties * globals, const char * envName, const char * propertyName, const char * defaultPrefix, const char * defaultSuffix)
```
Comment (afishbeck) [inline tools/esdlcmd-xml/esdl2xml.cpp]: is this the same as "extractEsdlCmdOption" below?  can they be commoned up?

File: tools/esdlcmd/esdlcmd_common.hpp [hunk 3]
```diff
@@ -103,7 +103,11 @@ typedef IEsdlCommand *(*EsdlCommandFactory)(const char *cmdname);
 #define ESDL_OPTION_ECL_INCLUDE_LIST    "--ecl-imports"
 #define ESDL_OPTION_ECL_HEADER_BLOCK    "--ecl-header"
 
-
+#define ESDLOPT_INCLUDE_PATH            "--includePath"
```
Comment (afishbeck) [inline tools/esdlcmd/esdlcmd_common.hpp]: in the name of consistency, eclcmd and esdlcmd normally use the format "--include-path".  Eclcc is a bit inconsistent.

File: tools/esdlcomp/esdlcomp.cpp [hunk 4]
```diff
@@ -1109,22 +1109,20 @@ ESDLcompiler::ESDLcompiler(const char * sourceFile, bool generatefile, const cha
     StringBuffer prot;
     splitFilename(sourceFile, &prot, &srcDir, &name, &ext);
 
+    if (includePath_ && *includePath_)
```
Comment (afishbeck) [inline tools/esdlcomp/esdlcomp.cpp]: Should you just go ahead and parse this into an array here?

File: tools/esdlcomp/esdlcomp.cpp [hunk 5]
```diff
@@ -1154,6 +1152,61 @@ ESDLcompiler::ESDLcompiler(const char * sourceFile, bool generatefile, const cha
...
+                    pathInInclude.append(PATHSEPCHAR);
+                pathInInclude.append(fname);
+                VStringBuffer pathInIncludeFull("%s%s", pathInInclude.str(), ext);
+                OwnedIFile fileInInclude = createIFile(pathInIncludeFull.str());
+                if (fileInInclude->exists())
+                {
+                    filepath.set(pathInIncludeFull.str());
+                    return true;
+                }
+                VStringBuffer altPathInIncludeFull("%s%s", pathInInclude.str(), alt_ext);
```
Comment (afishbeck) [inline tools/esdlcomp/esdlcomp.cpp]: You could reuse the existing variable and do 

pathInIncludeFull.setf("%s%s", pathInInclude.str(), alt_ext);

here.

File: tools/esdlcomp/esdlcomp.cpp [hunk 6]
```diff
@@ -1109,22 +1109,20 @@ ESDLcompiler::ESDLcompiler(const char * sourceFile, bool generatefile, const cha
...
-            StringBuffer alternateExtFilename;
-            alternateExtFilename.setf("%s%s%s", (prot.length()>0) ? prot.str() : "", srcDir.str(), name.str());
-
-            if (stricmp(ext.str(), ESDL_FILE_EXTENSION)==0)
-                alternateExtFilename.append(LEGACY_FILE_EXTENSION);
-            else
-                alternateExtFilename.append(ESDL_FILE_EXTENSION);
-
-            yyin = fopen(alternateExtFilename.str(), "rt");
+            StringBuffer includedFilePath;
```
Comment (afishbeck) [inline tools/esdlcomp/esdlcomp.cpp]: there's only a little 'd' between the names includeFilePath and includedFilePath.  Maybe this variable could be called "locatedFilePath"?

**Review by afishbeck — CHANGES_REQUESTED** [review]
@mayx a few comments.

**Comment by afishbeck** [comment]
@mayx plus the multiple -I support we discussed offline.

**Comment by mayx** [comment]
@afishbeck those are good suggestions, will make changes accordingly...

**Comment by mayx** [comment]
@afishbeck changes made, please review again...

File: tools/esdlcomp/esdlcomp.cpp (hunk 4 above)
Comment (afishbeck) [reply tools/esdlcomp/esdlcomp.cpp]: @mayx any comment on the idea of parsing this into a string array right away?

**Review by afishbeck — COMMENTED** [review]
.

**Review by afishbeck — APPROVED** [review]
@mayx looks good.  Not sure about the build errors?

**Review by afishbeck — APPROVED** [review]
@mayx looks good

**Pull request merged**